rust/src/delta.rs

```diff
@@ -660,7 +680,10 @@ impl DeltaTable {
 
     /// Updates the DeltaTable to the latest version by incrementally applying newer versions.
     /// It assumes that the table is already updated to the current version `self.version`.
-    pub async fn update_incremental(&mut self) -> Result<(), DeltaTableError> {
+    ///
+    /// It's private since it's expected for users to call `update` as it takes into account
+    /// checkpoints which handle expired tombstones.
+    async fn update_incremental(&mut self) -> Result<(), DeltaTableError> {
```

Comment:
It's unsafe to leave `update_incremental` as public, because it'll ignore the checkpoints which might have removed expired tombstones. Basically, since the `update` is fixed and optimized, it is safe to call it instead, e.g. it'll call `update_incremental` if no new checkpoint is found anyway

Reply:
Clarification question: does that mean we _have_ to use snapshots even though a sequence of commits may be substantially smaller, because only checkpoints capture the removal of tombstones?

My understanding is that for a typical query, the tombstone entries are not really applicable or relevant, as only the "live" files are required.

Reply:
I have the same question, this seems a bit off to me. `update` still has more overhead compared to `update_incremental` due to extra object store call to get the latest checkpoint and loading the checkpoint itself if it's one checkpoint behind.

Why do we care about whether an expired tombstone has been removed from checkpoint or not? If we are keeping tracking of expired tombstone correctly in memory, we shouldn't be getting any expired tombstone from `unexpired_tombstones` call. I think it's fine to include expired tombestone in `all_tombstones` call returns even if those tombstones are not included in the latest checkpoint.

Reply:
These changes was the easiest way to achieve the cleaning tombstones from state. Otherwise if we keep it as is and not address it, it'll overflow the memory. 
^ That's exactly how we found this in production, however from the writer perspective. 

But yeah, if additional call to get the latest checkpoints is overhead we could find other ways. Such as rely on the `CHECKPOINT_INTERVAL` for example to read the checkpoint or clear the state on update. I'll also try to find the spark reader behavior on this as I was focused primarily on a writer.

Reply:
I feel like spark does more than that, on each update call, they read every new file from latest checkpoint and if new checkpoint found then it's applied

Reply:
tbh, I'm open to any suggestion on this

Reply:
Maybe a read-only application could specify somehow that it's not interested in tombstones at all? IIUC, from a reader perspective the only use case for tombstones would be the time travel feature.

Depending on that config, the `update` method would be able decide if loading a checkpoint is a good idea.

Reply:
Ha, the tombstone GC problem is a good call. To me, it seems like the optimal setup would be to delete older in memory expired tombstones after an incremental update. I think there is value to have an incremental update code path for performance sensitive workloads and let users choose which one they want to use depending on the requirements.

Reply:
Letting read-only application specifying whether tombstones should be kept at all sounds like a neat trick to reduce memory usage :) I think it's better handled as a separate self-contained PR though because it's an optimization specific to a subset of readers, while the problem we are dealing where applies to both readers and writers.

Reply:
If needed, we could also fix tombstone gc for incremental updates in a separate PR since we already have `update` as a good workaround.

Reply:
I am having a second thought on @dispanser 's optimization idea, looks like it's more applicable than just a subset of readers? I don't think tombstones are required for time travels since during time travel call, we just start with a clean state and reload the table up to specified versions. I think it's also not required for writers that are not writing out checkpoints, which is the case for how we run KDI. So it is only needed for checkpoint and vacuum?

In order words, we could just drop all tombstones in KDI to further reduce memory usage if it's configured to use the external lambda checkpoint writer?

Reply:
@houqp - fyi: we are using internal checkpoints at this point in our kdi deployment.

at the same time, an in-memory removal of tombstones after a checkpoint doesn't seem very far out of reach. even so, if we leave `update_incremental` public in this PR to not force clients to use `update`, this would allow us to address our short term issue in kdi quickly and follow up with optimization in separate PR.

thoughts?

Reply:
Okay, processing all of the above, I'll leave update_incremental public as it was and we might address performance improvements within an another scope

Reply:
Created #425 to track the no-tombstones idea.